lisa/sut_orchestrator/libvirt/platform.py

```diff
@@ -0,0 +1,896 @@
+
+
+class BaseLibvirtPlatform(Platform):
+    _supported_features: List[Type[Feature]] = [
+        SerialConsole,
+    ]
+
+    def __init__(self, runbook: schema.Platform) -> None:
+        super().__init__(runbook=runbook)
+        self.libvirt_conn_str: str
+        self.qemu_platform_runbook: BaseLibvirtPlatformSchema
+        self.host_node: Node
+
+    @classmethod
+    def type_name(cls) -> str:
+        return ""
+
+    @classmethod
+    def supported_features(cls) -> List[Type[Feature]]:
+        return BaseLibvirtPlatform._supported_features
+
+    @classmethod
+    def platform_runbook_type(cls) -> type:
+        return BaseLibvirtPlatformSchema
+
+    @classmethod
+    def node_runbook_type(cls) -> type:
+        return BaseLibvirtNodeSchema
+
+    def _initialize(self, *args: Any, **kwargs: Any) -> None:
+        libvirt_events_thread.init()
+
+        self.qemu_platform_runbook = self.runbook.get_extended_runbook(
+            self.__platform_runbook_type(), type_name=type(self).type_name()
+        )
+
+    def _prepare_environment(self, environment: Environment, log: Logger) -> bool:
+        # Ensure environment log directory is created before connecting to any nodes.
+        _ = environment.log_path
+
+        if len(self.qemu_platform_runbook.hosts) > 1:
+            log.warning(
+                "Multiple hosts are currently not supported. "
+                "Only the first host will be used."
+            )
+
+        host = self.qemu_platform_runbook.hosts[0]
+        if host.is_remote():
+            assert host.address
+            if not host.username:
+                raise LisaException("Username must be provided for remote host")
+            if not host.private_key_file:
+                raise LisaException("Private key file must be provided for remote host")
+
+            self.host_node = RemoteNode(
+                runbook=schema.Node(name="qemu-host"),
+                index=-1,
+                logger_name="qemu-host",
+                base_part_path=environment.environment_part_path,
+                parent_logger=log,
+            )
+
+            self.host_node.set_connection_info(
+                address=host.address,
+                username=host.username,
+                private_key_file=host.private_key_file,
+            )
+        else:
+            self.host_node = local_node_connect(
+                name="qemu-host",
+                base_part_path=environment.environment_part_path,
+                parent_logger=log,
+            )
+
+        self._init_libvirt_conn_string()
+        self._configure_environment(environment, log)
+
+        with libvirt.open(self.libvirt_conn_str) as qemu_conn:
+            return self._configure_node_capabilities(environment, log, qemu_conn)
+
+    def _deploy_environment(self, environment: Environment, log: Logger) -> None:
+        self._deploy_nodes(environment, log)
+
+    def _delete_environment(self, environment: Environment, log: Logger) -> None:
```

Comment:
Because this class has subclasses, if a method is intended to be called in this class only, not inheritable, use `__` (private) before them and move them after the other `_` (protected) methods. It's in LISA code guidelines, and it's common practice in Python.